File: frontend/lib/api/communication.ts
import { api } from './client';

export type CommunicationType =
  | 'NOTIFICATION'
  | 'CONVERSATION_MESSAGE'
  | 'CLIENT_MESSAGE'
  | 'INTERNAL_NOTE'
  | 'OUTREACH_MESSAGE'
  | 'SYSTEM_ALERT'
  | 'SECURITY_ALERT'
  | 'WORKFLOW_UPDATE'
  | 'APPROVAL_REQUEST'
  | 'DOCUMENT_COMMENT';

export type NotificationPriority =
  | 'LOW'
  | 'NORMAL'
  | 'HIGH'
  | 'URGENT'
  | 'CRITICAL_SECURITY';

export type DeliveryChannel =
  | 'IN_APP'
  | 'EMAIL'
  | 'SMS'
  | 'WEBHOOK'
  | 'PUSH'
  | 'REALTIME_WS';

export type DeliveryStatus =
  | 'PENDING'
  | 'QUEUED'
  | 'SENT'
  | 'DELIVERED'
  | 'FAILED'
  | 'DEAD_LETTER'
  | 'SUPPRESSED'
  | 'SKIPPED_QUIET_HOURS';

export type InboxState =
  | 'ACTIVE'
  | 'READ'
  | 'ARCHIVED'
  | 'SNOOZED'
  | 'DELETED';

export type MessageVisibility =
  | 'CLIENT_VISIBLE'
  | 'INTERNAL';

export type ConversationType =
  | 'DIRECT'
  | 'GROUP'
  | 'PROJECT'
  | 'PROPOSAL'
  | 'CONTRACT'
  | 'SUPPORT_TICKET'
  | 'CHANGE_REQUEST';

export interface InboxItem {
  id: string;
  tenant_id: string;
  recipient_id: string;
  notification_id?: string;
  conversation_message_id?: string;
  item_type: CommunicationType;
  title: string;
  summary: string;
  priority: NotificationPriority;
  state: InboxState;
  is_read: boolean;
  is_starred: boolean;
  is_pinned: boolean;
  read_at?: string;
  snoozed_until?: string;
  action_url?: string;
  entity_type?: string;
  entity_id?: string;
  created_at: string;
  updated_at: string;
}

export interface InboxUnreadCount {
  unread_count: number;
  tenant_id: string;
  recipient_id: string;
}

export interface NotificationRecord {
  id: string;
  tenant_id: string;
  notification_type: CommunicationType;
  priority: NotificationPriority;
  title: string;
  body: string;
  source_module: string;
  source_event_id?: string;
  entity_type?: string;
  entity_id?: string;
  action_url?: string;
  actor_id?: string;
  idempotency_key?: string;
  metadata?: Record<string, any>;
  created_at: string;
}

export interface ConversationMember {
  id: string;
  conversation_id: string;
  user_id: string;
  role: string;
  joined_at: string;
  last_read_at?: string;
}

export interface ConversationAttachment {
  id: string;
  message_id: string;
  file_name: string;
  file_type: string;
  file_size_bytes: number;
  file_url: string;
  is_verified: boolean;
  created_at: string;
}

export interface ConversationMessage {
  id: string;
  conversation_id: string;
  sender_id: string;
  body: string;
  visibility: MessageVisibility;
  parent_id?: string;
  is_edited: boolean;
  edit_count: number;
  created_at: string;
  updated_at: string;
  attachments?: ConversationAttachment[];
}

export interface Conversation {
  id: string;
  tenant_id: string;
  title: string;
  conversation_type: ConversationType;
  entity_type?: string;
  entity_id?: string;
  created_by: string;
  is_closed: boolean;
  closed_at?: string;
  closed_by?: string;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface ConversationDetail {
  conversation: Conversation;
  members: ConversationMember[];
  messages: ConversationMessage[];
}

export interface NotificationPreference {
  id: string;
  tenant_id: string;
  user_id: string;
  notification_type: CommunicationType;
  in_app_enabled: boolean;
  email_enabled: boolean;
  sms_enabled: boolean;
  webhook_enabled: boolean;
  push_enabled: boolean;
  quiet_hours_start?: string;
  quiet_hours_end?: string;
  timezone: string;
  created_at: string;
  updated_at: string;
}

export interface DeliveryRecord {
  id: string;
  tenant_id: string;
  notification_id?: string;
  recipient_id: string;
  channel: DeliveryChannel;
  status: DeliveryStatus;
  provider?: string;
  provider_message_id?: string;
  retry_count: number;
  max_retries: number;
  next_retry_at?: string;
  sent_at?: string;
  delivered_at?: string;
  failed_at?: string;
  error_details?: Record<string, any>;
  created_at: string;
}

export const communicationApi = {
  // Inbox API
  listInbox: (params?: {
    state?: InboxState;
    is_read?: boolean;
    is_pinned?: boolean;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.state) query.append('state', params.state);
    if (params?.is_read !== undefined) query.append('is_read', String(params.is_read));
    if (params?.is_pinned !== undefined) query.append('is_pinned', String(params.is_pinned));
    if (params?.limit) query.append('limit', String(params.limit));
    if (params?.offset) query.append('offset', String(params.offset));
    const qs = query.toString();
    return api<InboxItem[]>(`/inbox${qs ? `?${qs}` : ''}`);
  },

  getUnreadCount: () => api<InboxUnreadCount>('/inbox/unread-count'),

  updateInboxItem: (
    itemId: string,
    payload: {
      state?: InboxState;
      is_read?: boolean;
      is_starred?: boolean;
      is_pinned?: boolean;
      snoozed_until?: string;
    }
  ) => api<InboxItem>(`/inbox/${itemId}`, { method: 'PATCH', body: JSON.stringify(payload) }),

  bulkUpdateInbox: (payload: {
    item_ids: string[];
    state?: InboxState;
    is_read?: boolean;
    is_starred?: boolean;
    is_pinned?: boolean;
    snoozed_until?: string;
  }) => api<{ updated_count: number; tenant_id: string; recipient_id: string }>('/inbox/bulk', {
    method: 'PATCH',
    body: JSON.stringify(payload),
  }),

  // Notifications API
  listNotifications: (params?: {
    source_module?: string;
    priority?: NotificationPriority;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.source_module) query.append('source_module', params.source_module);
    if (params?.priority) query.append('priority', params.priority);
    if (params?.limit) query.append('limit', String(params.limit));
    if (params?.offset) query.append('offset', String(params.offset));
    const qs = query.toString();
    return api<NotificationRecord[]>(`/notifications${qs ? `?${qs}` : ''}`);
  },

  createNotification: (payload: {
    notification_type: CommunicationType;
    priority?: NotificationPriority;
    title: string;
    body: string;
    source_module: string;
    source_event_id?: string;
    entity_type?: string;
    entity_id?: string;
    action_url?: string;
    recipients?: string[];
    channels?: DeliveryChannel[];
    idempotency_key?: string;
    metadata?: Record<string, any>;
  }) => api<any>('/notifications', { method: 'POST', body: JSON.stringify(payload) }),

  getNotification: (notificationId: string) =>
    api<NotificationRecord>(`/notifications/${notificationId}`),

  // Conversations API
  listConversations: (params?: {
    entity_type?: string;
    entity_id?: string;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.entity_type) query.append('entity_type', params.entity_type);
    if (params?.entity_id) query.append('entity_id', params.entity_id);
    if (params?.limit) query.append('limit', String(params.limit));
    if (params?.offset) query.append('offset', String(params.offset));
    const qs = query.toString();
    return api<Conversation[]>(`/conversations${qs ? `?${qs}` : ''}`);
  },

  createConversation: (payload: {
    title: string;
    conversation_type: ConversationType;
    entity_type?: string;
    entity_id?: string;
    initial_members?: string[];
    metadata?: Record<string, any>;
  }) => api<any>('/conversations', { method: 'POST', body: JSON.stringify(payload) }),

  getConversation: (conversationId: string) =>
    api<ConversationDetail>(`/conversations/${conversationId}`),

  postMessage: (
    conversationId: string,
    payload: {
      body: string;
      visibility?: MessageVisibility;
      parent_id?: string;
      attachments?: Array<{
        file_name: string;
        file_type: string;
        file_size_bytes: number;
        file_url: string;
      }>;
      metadata?: Record<string, any>;
    }
  ) => api<any>(`/conversations/${conversationId}/messages`, {
    method: 'POST',
    body: JSON.stringify(payload),
  }),

  editMessage: (
    messageId: string,
    payload: {
      new_body: string;
      edit_reason?: string;
    }
  ) => api<any>(`/conversations/messages/${messageId}`, {
    method: 'PATCH',
    body: JSON.stringify(payload),
  }),

  // Preferences API
  listPreferences: () => api<NotificationPreference[]>('/notification-preferences'),

  updatePreference: (payload: {
    notification_type: CommunicationType;
    in_app_enabled?: boolean;
    email_enabled?: boolean;
    sms_enabled?: boolean;
    webhook_enabled?: boolean;
    push_enabled?: boolean;
    quiet_hours_start?: string;
    quiet_hours_end?: string;
    timezone?: string;
  }) => api<NotificationPreference>('/notification-preferences', {
    method: 'PUT',
    body: JSON.stringify(payload),
  }),

  // Delivery Audit API
  listDeliveries: (params?: {
    status?: DeliveryStatus;
    channel?: DeliveryChannel;
    limit?: number;
    offset?: number;
  }) => {
    const query = new URLSearchParams();
    if (params?.status) query.append('status', params.status);
    if (params?.channel) query.append('channel', params.channel);
    if (params?.limit) query.append('limit', String(params.limit));
    if (params?.offset) query.append('offset', String(params.offset));
    const qs = query.toString();
    return api<DeliveryRecord[]>(`/communication/deliveries${qs ? `?${qs}` : ''}`);
  },

  retryDelivery: (deliveryId: string) =>
    api<any>('/communication/deliveries/retry', {
      method: 'POST',
      body: JSON.stringify({ delivery_id: deliveryId }),
    }),
};

File: backend/app/schemas/communication.py
"""Pydantic request and response schemas for Phase 37 Unified Communication Infrastructure."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID
from pydantic import BaseModel, ConfigDict, Field

from app.communication.base import (
    CommunicationType,
    ConversationType,
    DeliveryChannel,
    DeliveryStatus,
    InboxState,
    MessageVisibility,
    NotificationPriority,
)


# =============================================================================
# 1. Unified Inbox Schemas
# =============================================================================

class InboxItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    recipient_id: UUID
    notification_id: Optional[UUID] = None
    conversation_message_id: Optional[UUID] = None
    item_type: str
    title: str
    summary: str
    priority: NotificationPriority
    state: InboxState
    is_read: bool
    is_starred: bool
    is_pinned: bool
    read_at: Optional[datetime] = None
    snoozed_until: Optional[datetime] = None
    action_url: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime


class InboxUnreadCountResponse(BaseModel):
    unread_count: int
    tenant_id: UUID
    recipient_id: UUID


class InboxItemUpdateRequest(BaseModel):
    state: Optional[InboxState] = None
    is_read: Optional[bool] = None
    is_starred: Optional[bool] = None
    is_pinned: Optional[bool] = None
    snoozed_until: Optional[datetime] = None


class InboxBulkUpdateRequest(BaseModel):
    item_ids: List[UUID]
    state: Optional[InboxState] = None
    is_read: Optional[bool] = None
    is_starred: Optional[bool] = None
    is_pinned: Optional[bool] = None
    snoozed_until: Optional[datetime] = None


class BulkUpdateResponse(BaseModel):
    updated_count: int
    tenant_id: UUID
    recipient_id: UUID


# =============================================================================
# 2. Notifications & Preferences Schemas
# =============================================================================

class NotificationCreateRequest(BaseModel):
    notification_type: CommunicationType = CommunicationType.OPERATIONAL
    priority: NotificationPriority = NotificationPriority.NORMAL
    title: str
    body: str
    source_module: str
    source_event_id: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[UUID] = None
    action_url: Optional[str] = None
    recipients: Optional[List[str]] = None
    channels: Optional[List[DeliveryChannel]] = None
    idempotency_key: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class NotificationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    notification_type: CommunicationType
    priority: NotificationPriority
    title: str
    body: str
    source_module: str
    source_event_id: Optional[str] = None
    entity_type: Optional[str] = None
    entity_id: Optional[UUID] = None
    action_url: Optional[str] = None
    actor_id: Optional[UUID] = None
    idempotency_key: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime


class NotificationPreferenceResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    user_id: UUID
    notification_type: CommunicationType
    in_app_enabled: bool
    email_enabled: bool
    sms_enabled: bool
    webhook_enabled: bool
    push_enabled: bool
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    timezone: str
    created_at: datetime
    updated_at: datetime


class NotificationPreferenceUpdateRequest(BaseModel):
    notification_type: CommunicationType
    in_app_enabled: Optional[bool] = None
    email_enabled: Optional[bool] = None
    sms_enabled: Optional[bool] = None
    webhook_enabled: Optional[bool] = None
    push_enabled: Optional[bool] = None
    quiet_hours_start: Optional[str] = None
    quiet_hours_end: Optional[str] = None
    timezone: Optional[str] = None


# =============================================================================
# 3. Conversations & Messages Schemas
# =============================================================================

class ConversationCreateRequest(BaseModel):
    title: str
    conversation_type: ConversationType = ConversationType.PROJECT
    entity_type: Optional[str] = None
    entity_id: Optional[UUID] = None
    initial_members: Optional[List[str]] = None
    metadata: Optional[Dict[str, Any]] = None


class ConversationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    title: str
    conversation_type: ConversationType
    entity_type: Optional[str] = None
    entity_id: Optional[UUID] = None
    created_by: UUID
    is_closed: bool
    closed_at: Optional[datetime] = None
    closed_by: Optional[UUID] = None
    metadata: Optional[Dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class ConversationMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    conversation_id: UUID
    user_id: UUID
    role: str
    joined_at: datetime
    last_read_at: Optional[datetime] = None


class ConversationAttachmentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    message_id: UUID
    file_name: str
    file_type: str
    file_size_bytes: int
    file_url: str
    is_verified: bool
    created_at: datetime


class ConversationMessageResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    conversation_id: UUID
    sender_id: UUID
    body: str
    visibility: MessageVisibility
    parent_id: Optional[UUID] = None
    is_edited: bool
    edit_count: int
    created_at: datetime
    updated_at: datetime
    attachments: List[ConversationAttachmentResponse] = []


class ConversationDetailResponse(BaseModel):
    conversation: ConversationResponse
    members: List[ConversationMemberResponse] = []
    messages: List[ConversationMessageResponse] = []


class ConversationMessageCreateRequest(BaseModel):
    body: str
    visibility: MessageVisibility = MessageVisibility.CLIENT_VISIBLE
    parent_id: Optional[UUID] = None
    attachments: Optional[List[Dict[str, Any]]] = None
    metadata: Optional[Dict[str, Any]] = None


class ConversationMessageEditRequest(BaseModel):
    new_body: str
    edit_reason: Optional[str] = None


# =============================================================================
# 4. Deliveries & Audit Schemas
# =============================================================================

class DeliveryRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    tenant_id: UUID
    notification_id: Optional[UUID] = None
    recipient_id: UUID
    channel: DeliveryChannel
    status: DeliveryStatus
    provider: Optional[str] = None
    provider_message_id: Optional[str] = None
    retry_count: int
    max_retries: int
    next_retry_at: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    delivered_at: Optional[datetime] = None
    failed_at: Optional[datetime] = None
    error_details: Optional[Dict[str, Any]] = None
    created_at: datetime


class DeliveryRetryRequest(BaseModel):
    delivery_id: UUID


class RealtimeEventEnvelope(BaseModel):
    event_type: str
    channel: str
    tenant_id: UUID
    payload: Dict[str, Any]
    timestamp: datetime = Field(default_factory=lambda: datetime.now())
